### PR TITLE
Expand docs on Iterator::intersperse

### PR DESCRIPTION
**Pull request by lukaslueg**
Unstable feature in #79524. This expands on the docs to bring them more in line with how other methods of `Iterator` are demonstrated.

### CODE REVIEW

**Comment by lukaslueg** [comment]
r? @m-ou-se
as she reviewed the whole feature. Excuse my ignorance if this is unwanted.

File: library/core/src/iter/adapters/intersperse.rs [hunk 1]
```diff
@@ -151,17 +151,15 @@ where
...
+    (
+        lo.saturating_sub(next_is_elem as usize).saturating_add(lo),
+        hi.and_then(|hi| hi.saturating_sub(next_is_elem as usize).checked_add(hi)),
+    )
 }
 
 fn intersperse_fold<I, B, F, G>(
     mut iter: Peekable<I>,
-    init: B,
+    mut accum: B,
```
Comment (KodrAus) [inline library/core/src/iter/adapters/intersperse.rs:162]: I think the previous field name here might've been better, as it's consistent with `init` used in all other `Iterator` combinators.

**Review by KodrAus — COMMENTED** [review]
Thanks for working on this @lukaslueg! I just had one nitpicky thought, but adding examples is great.

File: library/core/src/iter/adapters/intersperse.rs (hunk 1 above)
Comment (lukaslueg) [reply library/core/src/iter/adapters/intersperse.rs]: Force-pushed it away :-)

File: library/core/src/iter/traits/iterator.rs [hunk 2]
```diff
@@ -606,14 +629,36 @@ pub trait Iterator {
...
+    /// ```
+    ///
+    /// `intersperse_with` can be used in situations where the separator needs
+    /// to be computed:
+    /// ```
+    /// #![feature(iter_intersperse)]
+    ///
     /// let src = ["Hello", "to", "all", "people", "!!"].iter().copied();
     ///
+    /// // The closure mutably borrows it's context to generate an item.
```
Comment (m-ou-se) [inline library/core/src/iter/traits/iterator.rs:653]:
```suggestion
    /// // The closure mutably borrows its context to generate an item.
```

**Review by m-ou-se — APPROVED** [review]
This looks great! Thanks!

`r=me` with the typo fixed:

**Pull request merged**